From b678b18061c772ebf3a830454d5a68f2fab2123c Mon Sep 17 00:00:00 2001
From: ZhangPeng <zhangpeng362@huawei.com>
Date: Mon, 22 May 2023 20:20:49 +0800
Subject: [PATCH 035/263] [Huawei] userswap: add handling of ZERO_PAGE

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I6CAIM

--------------------------------

We call follow_page() with FOLL_DUMP to handle ZERO_PAGE. Although
FOLL_DUMP is intended for get_dump_page(), it just so happens that its
special treatment of the ZERO_PAGE (returning an error instead of doing
get_page) suits uswap very well. If somehow an abnormal page has sneaked
into the range, we won't oops here.

Signed-off-by: ZhangPeng <zhangpeng362@huawei.com>
---
 mm/userswap.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/mm/userswap.c b/mm/userswap.c
index 8fb230e7e5c8..4884e0fc73b2 100644
--- a/mm/userswap.c
+++ b/mm/userswap.c
@@ -108,7 +108,7 @@ static unsigned long pages_can_be_swapped(struct mm_struct *mm,
 		 * follow_page will inc page ref, dec the ref after we remap
 		 * the page.
 		 */
-		page = follow_page(vma, addr, FOLL_GET);
+		page = follow_page(vma, addr, FOLL_GET | FOLL_DUMP);
 		if (IS_ERR_OR_NULL(page)) {
 			ret = -ENODEV;
 			goto out_err;
@@ -416,9 +416,9 @@ int mfill_atomic_pte_nocopy(struct mm_struct *mm,
 
 	if (!vma_uswap_compatible(src_vma))
 		return -EINVAL;
-
-	page = follow_page(src_vma, src_addr, FOLL_GET | FOLL_MIGRATION);
-	if (!page)
+	page = follow_page(src_vma, src_addr, FOLL_GET | FOLL_MIGRATION |
+			   FOLL_DUMP);
+	if (IS_ERR_OR_NULL(page))
 		return -ENODEV;
 
 	src_pmd = mm_find_pmd(mm, src_addr);
-- 
2.17.1

